PWGHF/TableProducer/candidateCreatorB0.cxx

```diff
@@ -140,119 +149,166 @@ struct HfCandidateCreatorB0 {
-      auto trackParCov0 = getTrackParCov(track0);
-      auto trackParCov1 = getTrackParCov(track1);
-      auto trackParCov2 = getTrackParCov(track2);
-
-      // reconstruct 3-prong secondary vertex (D±)
-      if (df3.process(trackParCov0, trackParCov1, trackParCov2) == 0) {
-        continue;
-      }
+    static int ncol = 0;
 
-      const auto& secondaryVertex = df3.getPCACandidate();
-      trackParCov0.propagateTo(secondaryVertex[0], bz);
-      trackParCov1.propagateTo(secondaryVertex[0], bz);
-      trackParCov2.propagateTo(secondaryVertex[0], bz);
-
-      // D∓ → π∓ K± π∓
-      array<float, 3> pVecpiK = {track0.px() + track1.px(), track0.py() + track1.py(), track0.pz() + track1.pz()};
-      array<float, 3> pVecD = {pVecpiK[0] + track2.px(), pVecpiK[1] + track2.py(), pVecpiK[2] + track2.pz()};
-      auto trackParCovPiK = o2::dataformats::V0(df3.getPCACandidatePos(), pVecpiK, df3.calcPCACovMatrixFlat(),
-                                                trackParCov0, trackParCov1, {0, 0}, {0, 0});
-      auto trackParCovD = o2::dataformats::V0(df3.getPCACandidatePos(), pVecD, df3.calcPCACovMatrixFlat(),
-                                              trackParCovPiK, trackParCov2, {0, 0}, {0, 0});
-
-      int index0D = track0.globalIndex();
-      int index1D = track1.globalIndex();
-      int index2D = track2.globalIndex();
-
-      // loop over pions
-      for (const auto& trackPion : tracksPion) {
-        // minimum pT selection
-        if (trackPion.pt() < ptPionMin || !isSelectedTrackDCA(trackPion)) {
-          continue;
+    for (const auto& collision : collisions) {
+      auto primaryVertex = getPrimaryVertex(collision);
+      auto covMatrixPV = primaryVertex.getCov();
+
+      if (ncol % 10000 == 0) {
+        LOG(info) << ncol << " collisions parsed";
+      }
+      ncol++;
+
+      auto thisCollId = collision.globalIndex();
+      auto candsDThisColl = candsD.sliceBy(candsDPerCollision, thisCollId);
+
+      for (const auto& candD : candsDThisColl) { // start loop over filtered D candidates indices as associated to this collision in candidateCreator3Prong.cxx
+        hMassDToPiKPi->Fill(invMassDplusToPiKPi(candD), candD.pt());
+        hPtD->Fill(candD.pt());
+        hCPAD->Fill(candD.cpa());
+
+        // track0 <-> pi, track1 <-> K, track2 <-> pi
+        auto track0 = candD.prong0_as<TracksWithSel>();
+        auto track1 = candD.prong1_as<TracksWithSel>();
+        auto track2 = candD.prong2_as<TracksWithSel>();
+        auto trackParCov0 = getTrackParCov(track0);
+        auto trackParCov1 = getTrackParCov(track1);
+        auto trackParCov2 = getTrackParCov(track2);
+
+        std::array<float, 3> pVec0; // = {track0.px(), track0.py(), track0.pz()};
+        std::array<float, 3> pVec1; // = {track1.px(), track1.py(), track1.pz()};
+        std::array<float, 3> pVec2; // = {track2.px(), track2.py(), track2.pz()};
+
+        o2::dataformats::DCA dca0;
+        // dca0.set(track0.dcaXY(), track0.dcaZ());
```

Comment:
I would keep the initialisation (it will change only if `track.collisionId() != thisCollId`, otherwise the default one is the good one)